cli/pkg/workspace/plugin/kubeconfig.go

```diff
@@ -132,187 +134,212 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
+	case o.Name == "":
 		defer func() {
 			if err == nil {
 				_, err = fmt.Fprintf(o.Out, "Note: 'kubectl ws' now matches 'cd' semantics: go to home workspace. 'kubectl ws -' to go back. 'kubectl ws .' to print current workspace.\n")
 			}
 		}()
 		fallthrough
 
-	case "~", home:
-		homeWorkspace, err := o.kcpClusterClient.Cluster(core.RootCluster.Path()).TenancyV1alpha1().Workspaces().Get(ctx, "~", metav1.GetOptions{})
+	case o.Name == "~" || o.Name == home || strings.HasPrefix(o.Name, "~:") || strings.HasPrefix(o.Name, home):
+		newServerHost, err := o.moveHome(ctx, home)
 		if err != nil {
 			return err
 		}
-		config, err := o.ClientConfig.ClientConfig()
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+		// move to 'my' root
+	case o.Name == ":":
+		newServerHost, err := o.moveCurrentRoot(ctx)
 		if err != nil {
 			return err
 		}
-		u, _, err := pluginhelpers.ParseClusterURL(config.Host)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+	case o.Name == ".":
+		cfg, err := o.ClientConfig.ClientConfig()
 		if err != nil {
-			return fmt.Errorf("current URL %q does not point to a workspace", config.Host)
+			return err
 		}
-		uh, err := url.Parse(homeWorkspace.Spec.URL)
+		return printCurrentWorkspace(o.Out, cfg.Host, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
+
+	case o.Name == "-":
+		newServerHost, err := o.movePrevious(ctx, currentContext)
 		if err != nil {
-			return fmt.Errorf("invalid home workspace URL %q: %w", homeWorkspace.Spec.URL, err)
+			return err
 		}
+		return printCurrentWorkspace(o.Out, newServerHost, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
 
-		// We keep the old host, and append homeworkspace url. This allows users to have front-proxy
-		// host to be different from the workspace host.
-		newServerHost = u.Scheme + "://" + path.Join(u.Host, uh.Path)
-
-	case ".":
-		cfg, err := o.ClientConfig.ClientConfig()
+	case strings.Contains(o.Name, ".."):
+		newServerHost, err := o.moveBackwards(ctx)
 		if err != nil {
 			return err
 		}
-		return currentWorkspace(o.Out, cfg.Host, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
 
-	default:
-		cluster := logicalcluster.NewPath(o.Name)
-		if !cluster.IsValid() {
-			return fmt.Errorf("invalid workspace name format: %s", o.Name)
-		}
-		config, err := o.ClientConfig.ClientConfig()
+		// root legacy move
+	case o.Name == core.RootCluster.String():
+		newServerHost, err := o.moveRootLegacy(ctx)
 		if err != nil {
 			return err
 		}
-		u, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+		// system is bit different. Usually when moving to system workspaces you
+		// need to have special access and "if workspace exists" checks fails.
+		// If merged to "absolute path check" code gets convoluted :/
+	case strings.HasPrefix(o.Name, ":system"):
```

Comment:
We have a constant for `system` somewhere I believe.

Also here `HasPrefix` seems fishy. We have to properly split by `":"`.

Reply:
Idea was not to trigger this path is somebody created `root:system` or `foo:bar:system` in any ways. So only way to access `system` from cli is to be explicit - `:system`. 
Basically don't want this code path to be triggered by mistake by users.

Reply:
My comment was about using string comparison. Convert it to a logical cluster path and operate on that.

Reply:
Updated but this does not make it prettier in my opinion :D